services/gundeck/src/Gundeck/Push/Native.hs

```diff
@@ -61,58 +61,90 @@ push m addrs = do
     Just chunkSize -> pooledMapConcurrentlyN_ chunkSize (push1 m) addrs
 
 push1 :: NativePush -> Address -> Gundeck ()
-push1 m a = do
-  e <- view awsEnv
-  r <- Aws.execute e $ publish m a
-  case r of
-    Success _ -> do
-      Log.debug $
-        field "user" (toByteString (a ^. addrUser))
-          ~~ field "notificationId" (toText (npNotificationid m))
-          ~~ Log.msg (val "Native push success")
-      view monitor >>= counterIncr (path "push.native.success")
-    Failure EndpointDisabled _ -> onDisabled
-    Failure PayloadTooLarge _ -> onPayloadTooLarge
-    Failure EndpointInvalid _ -> onInvalidEndpoint
-    Failure (PushException ex) _ -> do
-      logError a "Native push failed" ex
-      view monitor >>= counterIncr (path "push.native.errors")
+push1 = push1' 0
   where
-    onDisabled =
-      handleAny (logError a "Failed to cleanup disabled endpoint") $ do
-        Log.info $
-          field "user" (toByteString (a ^. addrUser))
-            ~~ field "arn" (toText (a ^. addrEndpoint))
-            ~~ field "cause" ("EndpointDisabled" :: Text)
-            ~~ msg (val "Removing disabled endpoint and token")
-        view monitor >>= counterIncr (path "push.native.disabled")
-        Data.delete (a ^. addrUser) (a ^. addrTransport) (a ^. addrApp) (a ^. addrToken)
-        onTokenRemoved
-        e <- view awsEnv
-        Aws.execute e (Aws.deleteEndpoint (a ^. addrEndpoint))
-    onPayloadTooLarge = do
-      view monitor >>= counterIncr (path "push.native.too_large")
-      Log.warn $
-        field "user" (toByteString (a ^. addrUser))
-          ~~ field "arn" (toText (a ^. addrEndpoint))
-          ~~ msg (val "Payload too large")
-    onInvalidEndpoint =
-      handleAny (logError a "Failed to cleanup orphaned push token") $ do
-        Log.warn $
-          field "user" (toByteString (a ^. addrUser))
-            ~~ field "arn" (toText (a ^. addrEndpoint))
-            ~~ field "cause" ("InvalidEndpoint" :: Text)
-            ~~ msg (val "Invalid ARN. Deleting orphaned push token")
-        view monitor >>= counterIncr (path "push.native.invalid")
-        Data.delete (a ^. addrUser) (a ^. addrTransport) (a ^. addrApp) (a ^. addrToken)
-        onTokenRemoved
-    onTokenRemoved = do
-      i <- mkNotificationId
-      let c = a ^. addrClient
-      let r = singleton (target (a ^. addrUser) & targetClients .~ [c])
-      let t = a ^. addrPushToken
-      let p = singletonPayload (PushRemove t)
-      Stream.add i r p =<< view (options . optSettings . setNotificationTTL)
+    push1' :: Int -> NativePush -> Address -> Gundeck ()
+    push1' n m a = do
+      if n > retryInvalidThreshold
```

Comment:
I would have to make `onPersistentlyInvalidEndpoint` a top-level function, that is, move it out of the `where` clause, since I would have to repeat it otherwise. The `if then else` doesn't look beautiful, but the other option would move quite some code where it wouldn't belong in my opinion. 🤔

Reply:
Sounds reasonable :+1: